### PR TITLE
[CARBONDATA-3200] No-Sort compaction

### PR DESCRIPTION
**Pull request by NamanRastogi**
When the data is loaded with SORT_SCOPE as NO_SORT, and done compaction upon, the data still remains unsorted. This does not affect much in query. The major purpose of compaction, is better pack the data and improve query performance.

Now, the expected behaviour of compaction is sort to the data, so that after compaction, query performance becomes better. The columns to sort upon are provided by SORT_COLUMNS.

The new compaction works as follows:
1. Do sorting on unsorted & restructured data and store in temporary files
2. Pick a row from those temporary files, and already sorted carbondata files, according to a comparator on sort_columns.
3. Write data to a new segment (similar to old compaction flow).
4. Repeat steps 2 & 3 until no more rows are left.

Dependency: PR 3014

 - [ ] Any interfaces changed?  -->  No
 - [ ] Any backward compatibility impacted?   -->  No
 - [ ] Document update required?  ->  No
 - [ ] Testing done
 - [ ] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA.

### CODE REVIEW

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/2254/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.3.2, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/10299/

**Comment by qiuchenjian** [comment]
Need it add a param to control whether to sort when no_sort?  
Maybe user only want to merge small segments to a big segment, slow data merging may be unaccepted

**Comment by NamanRastogi** [comment]
@qiuchenjian We dont need a parameter to control to control whether to sort the data or not while compacting. The sole purpose of compaction is to improve subsequent query performance, and even though the compaction gets slow, it is doing what is intended.

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.3.2, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/10325/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/2071/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/2276/

**Comment by ravipesala** [comment]
@NamanRastogi Lots of code is duplicated here, Please try to unify with other compactor processor to avoid the duplication.

**Comment by kumarvishal09** [comment]
@NamanRastogi You can merge HybridSorter into CompactionResultSortProcessor. For unsorted file it will be same flow, for sorted file you add one adapter(InMemorySortTempFileChunkHolder) on top of  RawResultIterator which will be inline with SortTempFileChunkHolder class so Interface will be same.  And In SingleThreadFinalMerger expose one method which will take List of Sorted RawResultIterator and add to record holder heap(PrirotyQueue). 
InMemorySortTempFileChunkHolder you have to convert Object[] to IntermediateSortTempRow in getRowMethod

**Comment by kumarvishal09** [comment]
@NamanRastogi please resolve conflicts and fix all failure

File: processing/src/main/java/org/apache/carbondata/processing/sort/sortdata/InMemorySortTempChunkHolder.java [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+  public void closeStream() {
+    rawResultIterator.close();
+  }
+
+  /* below method will be used to get the sort temp row
+   *
+   * @return row
+   */
+  public IntermediateSortTempRow getRow() {
+//    //TODO add code to get directly Object[] Instead Of CarbonRow Object
```
Comment (kumarvishal09) [inline processing/src/main/java/org/apache/carbondata/processing/sort/sortdata/InMemorySortTempChunkHolder.java]: remove this commented code

File: processing/src/main/java/org/apache/carbondata/processing/merger/CarbonCompactionUtil.java [hunk 2]
```diff
@@ -400,24 +417,53 @@ private static int getDimensionDefaultCardinality(CarbonDimension dimension) {
...
+      // it indicates it is a restructured block
+      if (tableLastUpdatedTime > dataFileFooter.getSchemaUpdatedTimeStamp()) {
+        return true;
       }
     }
-    return restructuredBlockExists;
+    return false;
   }
+
+  public static boolean isSorted(TaskBlockInfo taskBlockInfo) throws IOException {
```
Comment (kumarvishal09) [inline processing/src/main/java/org/apache/carbondata/processing/merger/CarbonCompactionUtil.java]: Use CarbonUtil.readMetadataFile(tableblockInfo) for reading file footer ...if carbondata file version is V1/V2 this code will not able to read the file footer

File: processing/src/main/java/org/apache/carbondata/processing/merger/CarbonCompactionUtil.java [hunk 3]
```diff
@@ -400,24 +417,53 @@ private static int getDimensionDefaultCardinality(CarbonDimension dimension) {
...
+
+    FileReader fileReader = FileFactory.getFileHolder(FileFactory.getFileType(filePath));
+    ByteBuffer buffer =
+        fileReader.readByteBuffer(FileFactory.getUpdatedFilePath(filePath), fileSize - 8, 8);
+    fileReader.finish();
+
+    CarbonFooterReaderV3 footerReader = new CarbonFooterReaderV3(filePath, buffer.getLong());
+    FileFooter3 footer = footerReader.readFooterVersion3();
+
+    return footer.isIs_sort();
```
Comment (kumarvishal09) [inline processing/src/main/java/org/apache/carbondata/processing/merger/CarbonCompactionUtil.java]: Now during compaction it will read file footer twice which will impact the compaction perfornace ....so for this expose get/set method in TableBlockInfo which will store filefooter....and in AbstractQueryExecutor.java add a  check if filefooter is present then no need to read again 
please find the code reference 
org/apache/carbondata/core/scan/executor/impl/AbstractQueryExecutor.java:217

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/2328/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/2120/

File: integration/spark-common/src/main/scala/org/apache/carbondata/spark/rdd/StreamHandoffRDD.scala [hunk 4]
```diff
@@ -118,7 +118,7 @@ class StreamHandoffRDD[K, V](
     CommonUtil.setTempStoreLocation(split.index, carbonLoadModel, true, false)
     // use CompactionResultSortProcessor to sort data dan write to columnar files
     val processor = prepareHandoffProcessor(carbonTable)
-    val status = processor.execute(iteratorList)
+    val status = processor.execute(iteratorList, null)
```
Comment (kumarvishal09) [inline integration/spark-common/src/main/scala/org/apache/carbondata/spark/rdd/StreamHandoffRDD.scala]: Please add comment why need to pass null

File: processing/src/main/java/org/apache/carbondata/processing/merger/CarbonCompactionExecutor.java [hunk 5]
```diff
@@ -105,10 +105,15 @@ public CarbonCompactionExecutor(Map<String, TaskBlockInfo> segmentMapping,
    *
    * @return List of Carbon iterators
```
Comment (kumarvishal09) [inline processing/src/main/java/org/apache/carbondata/processing/merger/CarbonCompactionExecutor.java]: add comment for map type what output this method will return

File: processing/src/main/java/org/apache/carbondata/processing/merger/CarbonCompactionUtil.java [hunk 6]
```diff
@@ -400,24 +417,53 @@ private static int getDimensionDefaultCardinality(CarbonDimension dimension) {
...
+        return true;
       }
-      if (restructuredBlockExists) {
-        break;
+    }
+
+    return false;
+  }
+
+  public static boolean isRestructured(List<DataFileFooter> listMetadata,
```
Comment (kumarvishal09) [inline processing/src/main/java/org/apache/carbondata/processing/merger/CarbonCompactionUtil.java]: Please add comments for all public method which you have exposed

File: processing/src/main/java/org/apache/carbondata/processing/merger/CompactionResultSortProcessor.java [hunk 7]
```diff
@@ -156,21 +159,22 @@ public CompactionResultSortProcessor(CarbonLoadModel carbonLoadModel, CarbonTabl
    * This method will iterate over the query result and convert it into a format compatible
    * for data loading
    *
-   * @param resultIteratorList
+   * @param unsortedResultIteratorList
```
Comment (kumarvishal09) [inline processing/src/main/java/org/apache/carbondata/processing/merger/CompactionResultSortProcessor.java]: update comment

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.3.2, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/10376/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/2123/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/2127/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/2333/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.3.2, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/10381/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/2334/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/2128/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.3.2, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/10382/

File: integration/spark-common/src/main/scala/org/apache/carbondata/spark/rdd/StreamHandoffRDD.scala (hunk 4 above)
Comment (NamanRastogi) [reply integration/spark-common/src/main/scala/org/apache/carbondata/spark/rdd/StreamHandoffRDD.scala]: Done.

File: processing/src/main/java/org/apache/carbondata/processing/merger/CarbonCompactionUtil.java (hunk 6 above)
Comment (NamanRastogi) [reply processing/src/main/java/org/apache/carbondata/processing/merger/CarbonCompactionUtil.java]: Done

File: processing/src/main/java/org/apache/carbondata/processing/merger/CompactionResultSortProcessor.java (hunk 7 above)
Comment (NamanRastogi) [reply processing/src/main/java/org/apache/carbondata/processing/merger/CompactionResultSortProcessor.java]: Done

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.3.2, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/10383/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/2335/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.3.2, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/10385/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/2337/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/2137/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.3.2, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/10391/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/2343/

File: integration/spark-common/src/main/scala/org/apache/carbondata/spark/rdd/CarbonMergerRDD.scala [hunk 8]
```diff
@@ -180,10 +180,10 @@ class CarbonMergerRDD[K, V](
         context.addTaskCompletionListener { _ =>
           close()
         }
-        try {
```
Comment (kumarvishal09) [inline integration/spark-common/src/main/scala/org/apache/carbondata/spark/rdd/CarbonMergerRDD.scala]: Revert this change

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/2138/

File: core/src/main/java/org/apache/carbondata/core/util/DataFileFooterConverterV3.java [hunk 9]
```diff
@@ -73,6 +73,11 @@ public DataFileFooter convertDataFileFooter(FileHeader fileHeader, FileFooter3 f
     dataFileFooter.setNumberOfRows(footer.getNum_rows());
     dataFileFooter.setSegmentInfo(getSegmentInfo(footer.getSegment_info()));
     dataFileFooter.setSchemaUpdatedTimeStamp(fileHeader.getTime_stamp());
+    if (footer.isSetIs_sort()) {
+      dataFileFooter.setIsSorted(footer.isIs_sort());
+    } else {
+      dataFileFooter.setIsSorted(false);
```
Comment (kumarvishal09) [inline core/src/main/java/org/apache/carbondata/core/util/DataFileFooterConverterV3.java]: for older file if you are setting false then use table sort scope

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/2344/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.3.2, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/10392/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/2355/

File: processing/src/main/java/org/apache/carbondata/processing/sort/sortdata/SingleThreadFinalSortFilesMerger.java [hunk 10]
```diff
@@ -114,6 +113,31 @@ public void startFinalMerge() throws CarbonDataWriterException {
...
+   */
+  public void addInMemoryRawResultIterator(List<RawResultIterator> sortedRawResultMergerList,
+      SegmentProperties segmentProperties, CarbonColumn[] noDicAndComplexColumns,
+      DataType[] measureDataType)
+      throws CarbonSortKeyAndGroupByException {
+    for (RawResultIterator rawResultIterator : sortedRawResultMergerList) {
+      InMemorySortTempChunkHolder inMemorySortTempChunkHolder =
+          new InMemorySortTempChunkHolder(rawResultIterator, segmentProperties,
+              noDicAndComplexColumns, sortParameters, measureDataType);
+      inMemorySortTempChunkHolder.readRow();
```
Comment (ravipesala) [inline processing/src/main/java/org/apache/carbondata/processing/sort/sortdata/SingleThreadFinalSortFilesMerger.java]: Don't need to check hasNext here before reading the row first time?

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/2146/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/2364/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.3.2, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/10403/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/2150/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.3.2, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/10405/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/2155/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.3.2, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/10409/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/2368/

**Comment by kumarvishal09** [comment]
retest this please

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/2166/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/2380/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.3.2, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/10422/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/2191/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/2407/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.3.2, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/10447/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/2428/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/2210/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.3.2, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/10466/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.3.2, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/10470/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/2215/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.3.2, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/10473/

**Comment by ravipesala** [comment]
LGTM

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/2457/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/2240/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.3.2, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/10496/

**Comment by kumarvishal09** [comment]
LGTM
